source/MainPanel.cpp

```diff
@@ -359,11 +361,29 @@ void MainPanel::ShowScanDialog(const ShipEvent &event)
 		out << "Your scanners cannot make any sense of this " + target->Noun() + "'s interior.";
 	else if(event.Type() & ShipEvent::SCAN_OUTFITS)
 	{
-		out << "This " + target->Noun() + " is equipped with:\n";
+		if(!target->Outfits().empty())
+			out << "This " + target->Noun() + " is equipped with:\n";
+		else
+			out << "This " + target->Noun() + " is not equipped with any outfits.\n";
+
+		// Split target->Outfits() into categories, then iterate over them in order.
+		auto comparator = CompareStringsByGivenOrder(GameData::Category(CategoryType::OUTFIT));
+		map<string, map<const Outfit *, int, CompareOutfitsByName>, CompareStringsByGivenOrder> outfitsByCategory(comparator);
 		for(const auto &it : target->Outfits())
-			if(it.first && it.second)
-				out << "\t" << it.second << " "
-					<< (it.second == 1 ? it.first->Name() : it.first->PluralName()) << "\n";
+			outfitsByCategory[it.first->Category()][it.first] = it.second;
+		for(const auto &it : outfitsByCategory)
+		{
+			if(it.second.empty())
+				continue;
+
+			// Print the category's name and outfits in it.
+			out << "\t" << (it.first.empty() ? "Unknown" : it.first) << "\n";
+			for(const auto &it2 : it.second)
+				if(it2.first && it2.second)
+					out << "\t\t" << it2.second << " "
+						<< (it2.second == 1 ? it2.first->Name() : it2.first->PluralName())
```

Comment:
This is going to affect the various missile launcher/pod/box outfits, too, like here (I scanned a random merchant Bastion):

![2022-01-15-215011_276x66_scrot](https://user-images.githubusercontent.com/24784687/149638822-d00a81ff-bc2e-4fe0-a95f-632785bbb7d9.png)

`Heavy Rockets` would normally sort after `Heavy Rocket Racks` because `'s' > ' '`, but here it sorts before because the sorting was done on the singular forms.

Reply:
Aside: In 4 out of 6 callsites of PluralName(), the caller calls either Name or PluralName in roughly the same way:

```grep
source/Engine.cpp:2152:                         + (amount == 1 ? outfit->Name() : outfit->PluralName()) + ".";
--
source/GameAction.cpp-56-               bool isSingle = (abs(count) == 1);
source/GameAction.cpp:57:               string nameWas = (isSingle ? outfit->Name() : outfit->PluralName());
--
source/MainPanel.cpp:355:                                       out << " " << (it.second == 1 ? it.first->Name(): it.first->PluralName()) << "\n";
--
source/MainPanel.cpp:384:                                               << (it2.second == 1 ? it2.first->Name() : it2.first->PluralName())
--
source/ShipInfoPanel.cpp-569-                   bool isSingular = (it.second == 1 || it.first->Get("installable") < 0.);
source/ShipInfoPanel.cpp:570:                   string name = (isSingular ? it.first->Name() : it.first->PluralName());
```

Maybe give Name an optional parameter, `Name(unsigned int N = 1)`, and have it return the plural name if N != 1?

Reply:
Done. With 90e89d5dee9233f9950c6598879b8d1c1c01330e:

![2022-01-15-221006_279x122_scrot](https://user-images.githubusercontent.com/24784687/149639252-060d9cbd-f8c2-4866-9814-6d3ec9ba227b.png)

That commit removes the original use of ByName, but that class is still used by MapOutfitterPanel :)